### PR TITLE
Aleksei/network in url

### PR DESCRIPTION
**Pull request by iambeone**
Closes #ISSUE

**Description:**

Initial fix to change network by url requests like /cosmos-hub-mainnet/validators

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by Bitcoinera** [comment]
Interesting. But why do you think we need this? What is the purpose?

We are already connecting to each network every time the user selects it & logs in...

**Comment by iambeone** [comment]
@Bitcoinera this one https://github.com/luniehq/lunie/issues/3213

File: src/components/common/AppMenu.vue [hunk 1]
```diff
@@ -33,8 +33,8 @@
       @click.native="signIn()"
     />
     <div class="app-menu-main">
-      <router-link
-        class="app-menu-item hide-xs"
+      <routerLinkNetwork
+        cclass="app-menu-item hide-xs"
```
Comment (jbibla) [inline src/components/common/AppMenu.vue]:
```suggestion
        class="app-menu-item hide-xs"
```

File: src/components/common/RouterLinkNetwork.vue [hunk 2]
```diff
@@ -0,0 +1,50 @@
+<template>
+  <router-link
+    :class="cclass"
```
Comment (jbibla) [inline src/components/common/RouterLinkNetwork.vue]: this is very strange

File: src/routes.js [hunk 3]
```diff
@@ -299,7 +221,92 @@ export default [
...
     path: `/feature-not-available/:feature`,
     component: () => import(`./components/common/PageFeatureNotAvailable`),
+    name: `fns`,
     props: true
   },
-  { path: `*`, component: () => import(`./components/common/Page404`) }
+  {
+    path: `/:networkId`,
+    component: () => import(`./components/common/setNetwork`),
+    children: [
```
Comment (jbibla) [inline src/routes.js]: nice.

File: src/routes.js [hunk 4]
```diff
@@ -299,7 +221,92 @@ export default [
   {
     path: `/feature-not-available/:feature`,
     component: () => import(`./components/common/PageFeatureNotAvailable`),
+    name: `fns`,
```
Comment (jbibla) [inline src/routes.js]: prefer not to use abbreviations

File: src/components/common/RouterLinkNetwork.vue (hunk 2 above)
Comment (iambeone) [reply src/components/common/RouterLinkNetwork.vue]: Could you explain? We can't use class property there

File: src/routes.js (hunk 4 above)
Comment (iambeone) [reply src/routes.js]: This is not needed, i forgot to remove

File: src/components/common/AppMenu.vue (hunk 1 above)
Comment (jbibla) [inline src/components/common/AppMenu.vue]: can you use `:class` here?

File: src/components/common/RouterLinkNetwork.vue [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    title: {
+      type: String,
+      default: null
+    }
+  },
+  computed: {
+    ...mapGetters([`network`]),
+    url() {
+      if (this.network.length) {
+        return `/${this.network}${this.to}`
```
Comment (jbibla) [inline src/components/common/RouterLinkNetwork.vue]: can we remove `-mainnet` from the link? so just `/terra/` or `/cosmos/`?

for testnets, we can leave `-testnet` in.

File: src/components/common/RouterLinkNetwork.vue (hunk 2 above)
Comment (jbibla) [reply src/components/common/RouterLinkNetwork.vue]: i guess i would rather use a different word or change it with a symbol or something. i'm not sure. 

`:class="class_"` 

all feels strange.

File: src/components/common/RouterLinkNetwork.vue (hunk 5 above)
Comment (Bitcoinera) [reply src/components/common/RouterLinkNetwork.vue]: I just left `cosmos-hub` and yes, `terra` for terra-mainnet. But I used slice, so you might not like it xD (please don't use the vomit emoji)

File: src/components/common/RouterLinkNetwork.vue (hunk 2 above)
Comment (Bitcoinera) [reply src/components/common/RouterLinkNetwork.vue]: I changed it for `routerClass`. What do you think?

Comment (iambeone) [reply src/components/common/RouterLinkNetwork.vue]: I don't know why i have chosen cclass for this variable, so any changes will be good)

File: src/components/common/RouterLinkNetwork.vue (hunk 5 above)
Comment (iambeone) [reply src/components/common/RouterLinkNetwork.vue]: Rewritten a bit

File: src/components/common/NetworkSetter.vue [hunk 6]
```diff
@@ -8,15 +8,10 @@ export default {
...
-        )
-      } else {
-        return networks.find(
-          network => network === this.$route.params.networkId.concat(`-mainnet`)
-        )
-      }
+      return networks.find(
+        network =>
+          network.replace(/-mainnet$/, ``) === this.$route.params.networkId
+      )
```
Comment (Bitcoinera) [inline src/components/common/NetworkSetter.vue]: Nice! Didn't know this method replace either. Great stuff

File: src/components/common/NetworkSetter.vue [hunk 7]
```diff
@@ -35,9 +30,13 @@ export default {
       /* istanbul ignore next */
       update(data) {
         const networks = data.networks.map(network => network.id)
-        if (this.$route.params.networkId && this.findNetwork(networks)) {
+        let networkTitle
+        if (
+          this.$route.params.networkId &&
+          (networkTitle = this.findNetwork(networks))
```
Comment (Bitcoinera) [inline src/components/common/NetworkSetter.vue]:
```suggestion
          (networkTitle === this.findNetwork(networks))
```

File: src/components/common/RouterLinkNetwork.vue [hunk 8]
```diff
@@ -36,9 +36,7 @@ export default {
     ...mapGetters([`network`]),
     url() {
       if (this.network.length) {
-        const networkName = this.network.includes(`mainnet`)
-          ? this.network.slice(0, -8)
-          : this.network
+        const networkName = this.network.replace(/-mainnet$/, ``)
```
Comment (Bitcoinera) [inline src/components/common/RouterLinkNetwork.vue]: Cool

File: src/components/common/NetworkSetter.vue (hunk 7 above)
Comment (Bitcoinera) [reply src/components/common/NetworkSetter.vue]: Ok, this actually works. I am learning so much looking at your code @iambeone !

File: src/routes.js [hunk 9]
```diff
@@ -301,5 +223,89 @@ export default [
     component: () => import(`./components/common/PageFeatureNotAvailable`),
     props: true
   },
-  { path: `*`, component: () => import(`./components/common/Page404`) }
+  {
+    path: `/:networkId`,
+    component: () => import(`./components/common/NetworkSetter`),
+    children: [
```
Comment (jbibla) [inline src/routes.js]: should we make session screens children of network?

Comment (Bitcoinera) [reply src/routes.js]: Doesn't make too much sense right? They are they same no matter which network you are connected to

File: src/components/common/NetworkSetter.vue (hunk 7 above)
Comment (faboweb) [reply src/components/common/NetworkSetter.vue]: this might work, but is confusing, this assignes networkTitle and returns true if networkTitle is not undefined. Let's make the code more readable by using findNetwork in line 33 and using a this.$route.params.networkId check inside of findNetwork

File: src/components/common/RouterLinkNetwork.vue (hunk 8 above)
Comment (faboweb) [reply src/components/common/RouterLinkNetwork.vue]: this forces us to have the same id schema for every network. :/

Comment (faboweb) [reply src/components/common/RouterLinkNetwork.vue]: I think we should avoid string manipulations on IDs wherever possible

File: src/components/common/RouterLinkNetwork.vue [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    },
+    title: {
+      type: String,
+      default: null
+    }
+  },
+  computed: {
+    ...mapGetters([`network`]),
+    url() {
+      if (this.network.length) {
```
Comment (faboweb) [inline src/components/common/RouterLinkNetwork.vue]: is `network` an array?

**Comment by faboweb** [comment]
Why not use a routeGuard as a solution? This way we need to implement the RouterLinkNetwork everywhere and new devs might forget.

File: src/components/common/RouterLinkNetwork.vue (hunk 8 above)
Comment (iambeone) [reply src/components/common/RouterLinkNetwork.vue]: > I think we should avoid string manipulations on IDs wherever possible

Why?

File: src/components/common/RouterLinkNetwork.vue (hunk 10 above)
Comment (iambeone) [reply src/components/common/RouterLinkNetwork.vue]: Nope, just checking if it's not empty

**Comment by iambeone** [comment]
> Why not use a routeGuard as a solution? This way we need to implement the RouterLinkNetwork everywhere and new devs might forget.

How we can use it? What should it replace?

File: src/components/common/RouterLinkNetwork.vue (hunk 8 above)
Comment (faboweb) [reply src/components/common/RouterLinkNetwork.vue]: because IDs are IDs and this forces us to a naming schema. IDs should not need to follow a naming schema.

File: src/components/common/RouterLinkNetwork.vue (hunk 10 above)
Comment (faboweb) [reply src/components/common/RouterLinkNetwork.vue]: well then not use `length` but check if it is empty. this code suggests it is an array...

File: src/components/common/RouterLinkNetwork.vue (hunk 8 above)
Comment (iambeone) [reply src/components/common/RouterLinkNetwork.vue]: But since then, how we are going to remove useless parts like mainnet?

Comment (faboweb) [reply src/components/common/RouterLinkNetwork.vue]: probably doing a lookup

**Comment by faboweb** [comment]
we already use route guards. check initializeApp line 39. there you can rewrite the request and prefix the network name if needed.

**Comment by faboweb** [comment]
but, let's ignore the centralization for now and take your solution for now. please follow up on the other comments though.

File: src/components/common/RouterLinkNetwork.vue (hunk 8 above)
Comment (Bitcoinera) [reply src/components/common/RouterLinkNetwork.vue]: I am not sure how to do this. 

If the ID doesn't follow any naming schema/pattern, then how can we know which is the lexeme of the network and which the "useless" parts.

Maybe I didn't understand you correctly, because I don't see any possible solution to this

Comment (faboweb) [reply src/components/common/RouterLinkNetwork.vue]: there should not be "parts" in an ID. it could be random. that is why I forced everyone to use the `testnet` variable from the networks response. This is the same case. If we want the URL to be different then the ID, we should do a dictionary lookup. Or we just pass the `route` from the API. which is preferable as we don't want network specific logic on the client.

**Comment by faboweb** [comment]
e2e tests failed

**Comment by Bitcoinera** [comment]
> e2e tests failed

When I push nearly 90% of the times I get "Max attempts reached" in the e2e tests

File: src/components/common/RouterLinkNetwork.vue (hunk 8 above)
Comment (iambeone) [reply src/components/common/RouterLinkNetwork.vue]: Maybe we can just store path in our db?

Comment (iambeone) [reply src/components/common/RouterLinkNetwork.vue]: Updated

**Comment by faboweb** [comment]
unit and e2e tests are failing

**Comment by Bitcoinera** [comment]
We should fix the e2e tests... they fail since they don't correctly wait for the netlified version to be deployed. It is a pity.

**Comment by faboweb** [comment]
They are failing in here because of this PR not because of netlify unavailability (that happens only on first push)

**Comment by Bitcoinera** [comment]
> They are failing in here because of this PR not because of netlify unavailability (that happens only on first push)

Weird, you are right that the netlified version is already there. Why is it failing with "Max attempts reached" though?

**Comment by faboweb** [comment]
Yep, it'S actually a bug ^^'

**Comment by faboweb** [comment]
I confirm this is not working

File: src/components/common/RouterLinkNetwork.vue (hunk 8 above)
Comment (jbibla) [reply src/components/common/RouterLinkNetwork.vue]: #slugs

File: src/routes.js (hunk 9 above)
Comment (jbibla) [reply src/routes.js]: my thinking is that it would simpler because the user wouldn't have to choose a network before a session flow. 

we could also share links to create / recovering / signing in / exploring for specific networks which may be valuable for support questions and for customers who want to share a link to creating an address on their own network.

... just a thought. doesn't have to be done here / now.

**Comment by iambeone** [comment]
Fixed some redirect issues and rewrite to routeGuard, but e2e tests still need to be fixed. Working on it

File: src/routes.js [hunk 11]
```diff
@@ -6,306 +7,312 @@
...
-    name: `welcome`,
-    components: {
-      session: () => import(`./components/common/TmSessionWelcome`)
+export default (apollo, store) => {
+  return [
+    {
+      path: `/`,
+      beforeEnter: (to, from, next) =>
+        setNetwork({ to, from, next }, apollo, store),
+      component: () => import(`./components/common/NetworkSetter`)
```
Comment (faboweb) [inline src/routes.js]: This component doesn't exist anymore

File: src/routes.js [hunk 12]
```diff
@@ -6,306 +7,312 @@
...
+      component: () => import(`./components/common/PageCareers`)
+    },
+    {
+      path: `/feature-not-available/:feature`,
+      component: () => import(`./components/common/PageFeatureNotAvailable`),
+      props: true
+    },
+    {
+      path: `/:networkId`,
+      component: () => import(`./components/common/NetworkSetter`),
```
Comment (faboweb) [inline src/routes.js]: same

File: src/scripts/setNetwork.js [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    fetchPolicy: "cache-first"
+  })
+  let path = to.path
+  if (path == "/") {
+    path = "/portfolio"
+  }
+  // current network slug
+  let network
+  if (
+    store.state.connection.networkSlug != to.params.networkId &&
```
Comment (faboweb) [inline src/scripts/setNetwork.js]:
```suggestion
    store.state.connection.networkSlug !== to.params.networkId &&
```

File: src/scripts/setNetwork.js [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    network = networks.find(network => network.slug === to.params.networkId)
+    if (!network) {
+      network = networks.find(network => network.default === true)
+    }
+    store.dispatch(`setNetwork`, network)
+    next(`/${network.slug}${path}`)
+  } else if (!to.params.networkId) {
+    // swithing to current network
+    if (store.state.connection.networkSlug) {
+      network = networks.find(
```
Comment (faboweb) [inline src/scripts/setNetwork.js]: these lines are doublicated with lines 30+, refactor

File: src/scripts/setNetwork.js [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        network => network.slug === store.state.connection.networkSlug
+      )
+      if (!network) {
+        network = networks.find(network => network.default === true)
+      }
+      store.dispatch(`setNetwork`, network)
+      next(`/${network.slug}${path}`)
+    } else {
+      network = networks.find(network => network.default === true)
+      next(`/${network.slug}${path}`)
```
Comment (faboweb) [inline src/scripts/setNetwork.js]: why is there no store.dispatch(`setNetwork`, network) in this case?

**Comment by faboweb** [comment]
Looking good! Smaller issues.

File: src/routes.js (hunk 11 above)
Comment (Bitcoinera) [reply src/routes.js]: ? it does still exist. And still necessary it seems. It is located in `src/components/common`

File: src/routes.js (hunk 12 above)
Comment (Bitcoinera) [reply src/routes.js]: it's still alive! xD

File: src/scripts/setNetwork.js (hunk 13 above)
Comment (Bitcoinera) [reply src/scripts/setNetwork.js]: Done

File: src/scripts/setNetwork.js (hunk 14 above)
Comment (Bitcoinera) [reply src/scripts/setNetwork.js]: Refactored

File: src/scripts/setNetwork.js (hunk 15 above)
Comment (Bitcoinera) [reply src/scripts/setNetwork.js]: Yes, I bad fixed it. Now it's good again

**Comment by Bitcoinera** [comment]
> Looking good! Smaller issues.

Not so small.

Things currently failing:

- The `Back to validators` button still goes to the old url, `/validators`, thus failing.

- When you sign in with extension, we get the old bug of being always connected to Cosmos Hub 3, and thus failing too.

- I had to do a strange hack to pass the unit tests. Concretely the one about the fallback network. You can see my hack at line 54 of `vuex/modules/connection.js`. Otherwise it wouldn't get inside the `setNetwork` function 🤷‍♀  Maybe it's fine like this. It seems to be working this way.

File: src/vuex/modules/connection.js [hunk 16]
```diff
@@ -34,14 +34,14 @@ export default function({ apollo }) {
         ? data.networks.find(network => network.id === persistedNetwork)
         : false
       if (persistedNetwork && storedNetwork) {
-        await commit(`setNetworkId`, storedNetwork.id)
+        await dispatch(`setNetwork`, storedNetwork)
```
Comment (faboweb) [inline src/vuex/modules/connection.js]: why was this necessary?

Comment (Bitcoinera) [reply src/vuex/modules/connection.js]: Because this is the proper way, otherwise it fails in execution. But I changed it yesterday fast to pass the tests 😅

**Comment by faboweb** [comment]
we can simplify this more if we always use the network slug for all routes I belive

File: src/scripts/setNetwork.js [hunk 17]
```diff
@@ -20,11 +20,33 @@ export const setNetwork = async ({ to, next }, apollo, store) => {
...
+  if (to.params.networkId) {
+    const subPathRegexp = /\/.+(\/.+)/ //extracts the subpath
+    const match = subPathRegexp.exec(to.path)
+    if (match && match.length >= 2) {
+      // the path is is of form /cosmos-hub/portfolio
+      path = match[1]
+    } else {
+      // the path is is of form /cosmos-hub so we forward to portfolio
+      path = "/portfolio"
+    }
```
Comment (Bitcoinera) [inline src/scripts/setNetwork.js]: There is something here I don't get. Wasn't the reloading happening in production?

Or it was also happening here? In what cases? I didn't notice 👀

**Comment by Bitcoinera** [comment]
> we can simplify this more if we always use the network slug for all routes I belive

Also the session ones right? And then just centralize it

File: src/routes.js (hunk 11 above)
Comment (faboweb) [reply src/routes.js]: this is useless, need to remove
